File: src/doubly_even/enumerate/quotient.py
"""Q_C-coordinate candidate pipeline (Milestone 4 phase (a)).

For a doubly even code ``C`` of length ``N`` and rank ``k`` we want the
``Aut(C)``-orbit reps of nonzero cosets ``v + C ⊆ C⊥`` with
``wt(v) ≡ 0 (mod 4)`` — the valid doubly even augmentations.

The reference pipeline in :mod:`.filters` does this in ``F_2^N``: it
enumerates the ``2^(n-2k)`` canonical coset reps via
``standard_form_coset_reps``, filters by ``wt mod 4``, then runs an
orbit-min BFS that per step *applies* a column permutation and then
*reduces mod C*.

This module does the same work, but with the cosets re-coordinatised as
``(n - 2k)``-bit integers in a fixed basis of
``V := C⊥ ∩ {v : v_p = 0 ∀ p ∈ pivots(C)}`` (the canonical-rep
subspace). The BFS body becomes one XOR per set bit of the current
element, with no reduction step — the quotient is built into the
coordinates. We then lift orbit-min survivors back to ``F_2^N`` and
apply the ``wt mod 4`` filter on the lift.

Math note. The plan doc framed this in terms of the quadratic form
``q(v + C) := wt(v)/2 mod 2`` and claimed its polarisation is zero
because ``⟨u, v⟩ ≡ 0 (mod 2)`` for all ``u, v ∈ C⊥``. The second
claim only holds when ``C`` is self-dual; for ``C ⊊ C⊥`` the
polarisation is generally nonzero, so a single ``q_vector & u`` bit
parity does not classify singular vectors. We therefore keep the
weight check at the lift step rather than encoding ``q`` in
``Q``-coordinates. Lifting is one XOR per set bit and runs only on
the orbit-min survivor list, so the cost is negligible.

The output type of :func:`doubly_even_candidates_Q` matches the
existing ``doubly_even_candidates``: a sorted ``list[int]`` of
``F_2^N`` vectors. ``filters.doubly_even_candidates`` delegates here.
"""

from __future__ import annotations

from collections.abc import Iterable

from ..spec.codes import Code
from ..spec.vectors import wt


def Q_basis(C: Code) -> tuple[tuple[int, ...], tuple[int, ...]]:
    """Return ``(V_basis, pivots_V)`` for ``Q_C := C⊥/C``.

    ``V_basis``: tuple of ``L = n - 2k`` integers in ``F_2^N``, each in
    ``C⊥`` and zero on every pivot column of ``C``. The tuple is
    row-reduced; ``pivots_V[i]`` is the column of ``V_basis[i]``'s
    leading 1.

    Precondition: ``C ⊆ C⊥`` (always holds on the doubly even
    augmentation tree).
    """
    rref_rows, pivots = C.rref_basis()
    dual_basis = C.dual().basis

    # Reduce each dual basis vector mod C — clears the bits at C's
    # pivot columns, sending each dual generator to V.
    reduced: list[int] = []
    for v in dual_basis:
        rep = v
        for row, p in zip(rref_rows, pivots):
            if (rep >> p) & 1:
                rep ^= row
        if rep != 0:
            reduced.append(rep)

    # ``dual.basis`` has n - k generators; ``dim V = n - 2k``. Row-reduce
    # to recover a basis of V and its pivot columns (which live in
    # ``[0, N) \ pivots(C)``).
    V_basis, pivots_V = Code(C.n, tuple(reduced)).rref_basis()
    return V_basis, pivots_V


def lift(u_Q: int, V_basis: tuple[int, ...]) -> int:
    """Map a ``Q``-coordinate int back to its ``F_2^N`` rep.

    ``u_Q`` is an ``L``-bit integer with ``L = len(V_basis)``. The
    output is the XOR of ``V_basis[i]`` over set bits ``i`` of ``u_Q``.
    """
    v = 0
    u = u_Q
    i = 0
    while u:
        if u & 1:
            v ^= V_basis[i]
        u >>= 1
        i += 1
    return v


def project(v_in_V: int, pivots_V: tuple[int, ...]) -> int:
    """Map an ``F_2^N`` vector already in ``V`` to its ``Q``-coordinate.

    ``V_basis`` is in RREF, so bit ``i`` of the output is the bit of
    ``v_in_V`` at column ``pivots_V[i]``. Callers must ensure
    ``v_in_V ∈ V`` (i.e. is zero on ``C``'s pivot columns); the
    function does not check.
    """
    out = 0
    for i, p in enumerate(pivots_V):
        if (v_in_V >> p) & 1:
            out |= 1 << i
    return out


def aut_image_on_Q(
    aut_generators: Iterable[tuple[int, ...]],
    C: Code,
    V_basis: tuple[int, ...],
    pivots_V: tuple[int, ...],
) -> list[tuple[int, ...]]:
    """Image of each automorphism generator in ``End(Q_C)``.

    For each ``σ`` in ``aut_generators``, returns a length-``L`` tuple
    ``sigma_Q`` where ``sigma_Q[i]`` is the ``Q``-coordinate of
    ``σ(V_basis[i])`` reduced mod ``C``. Applying ``sigma_Q`` to a
    ``Q``-coordinate int ``u`` is then ``XOR sigma_Q[i] for each bit i
    set in u``.
    """
    rref_rows, pivots = C.rref_basis()
    rref_pivots = list(zip(rref_rows, pivots))

    out: list[tuple[int, ...]] = []
    for sigma in aut_generators:
        sigma_list = list(sigma)
        # Image of each Q-basis vector.
        images: list[int] = []
        for b in V_basis:
            # Apply column permutation σ.
            permuted = 0
            for i, j in enumerate(sigma_list):
                if (b >> i) & 1:
                    permuted |= 1 << j
            # Reduce mod C (zero out bits at C's pivot columns).
            for row, p in rref_pivots:
                if (permuted >> p) & 1:
                    permuted ^= row
            # Project to Q-coordinates by reading V's pivot bits.
            coord = 0
            for k, p in enumerate(pivots_V):
                if (permuted >> p) & 1:
                    coord |= 1 << k
            images.append(coord)
        out.append(tuple(images))
    return out


def _sigma_Q_table(sigma_Q: tuple[int, ...]) -> list[int]:
    """Precompute ``table[u] = σ_Q · u`` for every ``u ∈ [0, 2^L)``.

    Built in Gray-code order so each entry costs one XOR vs. the
    previous: total build cost is ``2^L`` XORs (vs. ``2^L · L`` for
    the naïve per-cell computation). The BFS then applies ``σ_Q`` via
    a single list index, eliminating the per-step Python bit-walk
    that previously dominated the profile.
    """
    L = len(sigma_Q)
    size = 1 << L
    table = [0] * size
    if L == 0:
        return table
    val = 0
    u = 0
    for i in range(1, size):
        flip = (i & -i).bit_length() - 1
        u ^= 1 << flip
        val ^= sigma_Q[flip]
        table[u] = val
    return table


def aut_orbit_minima_Q(
    reps_Q: Iterable[int],
    sigma_Qs: list[tuple[int, ...]],
) -> list[int]:
    """One ``Q``-coordinate rep per ``⟨sigma_Qs⟩``-orbit on ``reps_Q``.

    Algorithm: decompose ``reps_Q`` into orbits in a single sweep. Each
    coset is visited exactly once across all orbits (vs. the naïve
    per-rep ``_is_orbit_min`` BFS which visits each orbit element up to
    ``orbit_size`` times). We iterate ``reps_Q`` in sorted order and
    BFS-expand the orbit of the first unseen element, yielding that
    element as the orbit min.

    Inner step: ``σ_Q · current`` is a single list index into a
    precomputed table (see :func:`_sigma_Q_table`). Memory cost is
    ``2^L`` ints per generator per call.

    Precondition for correctness of the "min in iteration order = orbit
    min" trick: ``reps_Q`` must be closed under the action (so each
    orbit is fully inside it). The caller satisfies this by passing
    only ``wt mod 4 = 0`` cosets, which is closed because ``Aut(C)``
    preserves ``wt mod 4`` on cosets of doubly even ``C``.

    With an empty generator set every rep is its own orbit — return a
    sorted list of the input.
    """
    if not sigma_Qs:
        return sorted(reps_Q)

    tables = [_sigma_Q_table(tuple(s)) for s in sigma_Qs]
    reps_sorted = sorted(reps_Q)
    seen: set[int] = set()
    minima: list[int] = []

    for v in reps_sorted:
        if v in seen:
            continue
        minima.append(v)
        seen.add(v)
        queue: list[int] = [v]
        while queue:
            next_queue: list[int] = []
            for current in queue:
                for table in tables:
                    new_v = table[current]
                    if new_v not in seen:
                        seen.add(new_v)
                        next_queue.append(new_v)
            queue = next_queue
    return minima


def singular_reps_Q(V_basis: tuple[int, ...]) -> list[int]:
    """Yield ``Q``-coordinate ints ``u`` whose lift has ``wt ≡ 0 (mod 4)``.

    Gray-code walk through ``[1, 2^L)`` maintaining ``(u, lift(u))``
    incrementally: each step toggles one ``Q``-bit and XORs the
    corresponding ``V_basis`` row into the lift. We keep ``u`` whenever
    the lift's popcount is divisible by 4 — this is exactly the singular
    coset condition.

    Returned as a list (the caller iterates it twice: once for orbit-min
    BFS, once via the lift step that sorts the survivors).
    """
    L = len(V_basis)
    out: list[int] = []
    if L == 0:
        return out
    u = 0
    v = 0
    for i in range(1, 1 << L):
        flip = (i & -i).bit_length() - 1
        u ^= 1 << flip
        v ^= V_basis[flip]
        if v.bit_count() & 3 == 0:
            out.append(u)
    return out


def doubly_even_candidates_Q(
    C: Code, aut_generators: Iterable[tuple[int, ...]]
) -> list[int]:
    """``Aut(C)``-orbit reps of doubly even 1-dim extensions of ``C``.

    Drop-in replacement for
    :func:`doubly_even.enumerate.filters.doubly_even_candidates` —
    returns the same sorted ``list[int]`` of ``F_2^N`` representatives.

    Pipeline:

    1. Build ``V_basis, pivots_V = Q_basis(C)``.
    2. Image each ``Aut(C)`` generator into ``End(Q_C)``.
    3. Enumerate the singular ``Q``-coordinates (lift weight ≡ 0 mod 4)
       via a Gray-code walk that maintains the lift incrementally.
    4. Orbit-min filter in ``Q``.
    5. Lift each survivor back to ``F_2^N``.

    The weight filter must run *before* orbit-min (not after) so the
    orbit BFS only visits the singular cosets — otherwise we'd burn
    BFS time on the ~3/4 of cosets that fail ``wt mod 4`` and never
    yield an output. ``Aut(C)`` preserves ``wt mod 4`` on cosets
    (column permutation preserves weight; coset reduction by ``c ∈ C``
    doubly even shifts weight by ``wt(c) - 2⟨w, c⟩ ≡ 0 (mod 4)`` for
    ``w ∈ C⊥``), so every orbit lies entirely inside or entirely
    outside the singular set — pre-filtering does not split orbits.
    """
    V_basis, pivots_V = Q_basis(C)
    sigma_Qs = aut_image_on_Q(aut_generators, C, V_basis, pivots_V)
    reps_Q = singular_reps_Q(V_basis)
    orbit_min = aut_orbit_minima_Q(reps_Q, sigma_Qs)
    return sorted(lift(u, V_basis) for u in orbit_min)

File: rust/core/src/quotient.rs
//! `Q_C := C⊥ / C` operations.
//!
//! Direct port of `doubly_even.enumerate.quotient.{Q_basis, lift, project,
//! aut_image_on_Q}`. The kernel receives `C`'s RREF basis + pivots and
//! `C.dual().basis` from the Python side (Python's
//! `Code.rref_basis()` / `Code.dual()` are the readable reference); the
//! kernel re-derives `V_basis` and `pivots_V` so callers don't have to
//! marshal them.

use crate::linalg::{apply_permutation, row_reduce};
use crate::types::{BinVec, ColPerm, Mat};

/// Reduce a vector `v` modulo a code given by `(rref, pivots)`.
///
/// Clears every bit of `v` at a pivot column of `C`. The result is the
/// unique coset representative of `v + C` whose bits at `C`'s pivots are
/// all zero.
///
/// Inlined hot helper — used by [`q_basis`] and [`aut_image_on_q`].
#[inline]
pub fn reduce_mod_c(mut v: BinVec, rref: &[BinVec], pivots: &[u32]) -> BinVec {
    for (row, &p) in rref.iter().zip(pivots.iter()) {
        if (v >> p) & 1 == 1 {
            v ^= *row;
        }
    }
    v
}

/// Return `(V_basis, pivots_V)` for `Q_C = C⊥ / C`.
///
/// `V_basis` is the RREF basis of the canonical-rep subspace
/// `V = C⊥ ∩ {v : v_p = 0 ∀ p ∈ pivots(C)}`; `pivots_V[i]` is the leading-1
/// column of `V_basis[i]`. `pivots_V[i] ∈ [0, n) \ pivots(C)` by construction.
///
/// Precondition: `C ⊆ C⊥` (always holds on the doubly even augmentation tree).
pub fn q_basis(
    rref: &[BinVec],
    pivots: &[u32],
    dual_basis: &[BinVec],
    n: u32,
) -> (Vec<BinVec>, Vec<u32>) {
    let mut reduced: Vec<BinVec> = Vec::with_capacity(dual_basis.len());
    for &v in dual_basis {
        let rep = reduce_mod_c(v, rref, pivots);
        if rep != 0 {
            reduced.push(rep);
        }
    }
    row_reduce(&reduced, n)
}

/// Map a `Q`-coordinate `u_q` back to an `F_2^N` representative.
///
/// `out = XOR_{i: bit i of u_q is set} V_basis[i]`.
#[inline]
pub fn lift(u_q: BinVec, v_basis: &[BinVec]) -> BinVec {
    let mut out: BinVec = 0;
    let mut u = u_q;
    while u != 0 {
        let i = u.trailing_zeros() as usize;
        out ^= v_basis[i];
        u &= u - 1;
    }
    out
}

/// Project an `F_2^N` vector already lying in `V` to its `Q`-coordinate.
///
/// `V_basis` is in RREF, so bit `i` of the output is the bit of `v_in_v` at
/// column `pivots_v[i]`. Callers must guarantee `v_in_v ∈ V`; we don't check.
#[inline]
pub fn project(v_in_v: BinVec, pivots_v: &[u32]) -> BinVec {
    let mut out: BinVec = 0;
    for (i, &p) in pivots_v.iter().enumerate() {
        if (v_in_v >> p) & 1 == 1 {
            out |= 1u64 << i;
        }
    }
    out
}

/// Image of each automorphism generator in `End(Q_C)`.
///
/// For each `σ` in `aut_generators` returns a length-`L` column-form matrix
/// `sigma_q` where `sigma_q[i]` is the `Q`-coordinate of `σ(V_basis[i])`
/// reduced mod `C`. Applying `sigma_q` to a `Q`-coordinate `u` is then
/// `XOR_{i: u_i = 1} sigma_q[i]`.
pub fn aut_image_on_q(
    aut_generators: &[ColPerm],
    rref: &[BinVec],
    pivots: &[u32],
    v_basis: &[BinVec],
    pivots_v: &[u32],
) -> Vec<Mat> {
    let mut out: Vec<Mat> = Vec::with_capacity(aut_generators.len());
    let l = v_basis.len();
    for sigma in aut_generators {
        let mut sigma_q: Mat = Vec::with_capacity(l);
        for &b in v_basis {
            let permuted = apply_permutation(b, sigma);
            let reduced = reduce_mod_c(permuted, rref, pivots);
            sigma_q.push(project(reduced, pivots_v));
        }
        out.push(sigma_q);
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Round-trip: `project(lift(u, V_basis), pivots_V) == u`.
    #[test]
    fn project_lift_round_trip() {
        // Any RREF basis with strictly increasing pivots works here.
        // We pin a small one: dim 3 in F_2^5, pivots at columns 0, 2, 3.
        let v_basis: Vec<BinVec> = vec![
            0b00001, // bit 0
            0b00100, // bit 2
            0b01000, // bit 3
        ];
        let pivots_v: Vec<u32> = vec![0, 2, 3];
        for u in 0u64..(1 << 3) {
            let lifted = lift(u, &v_basis);
            let reprojected = project(lifted, &pivots_v);
            assert_eq!(reprojected, u, "round trip failed at u={u:b}");
        }
    }

    /// `lift(0, _) == 0` and `lift(u, single) == single` when popcount(u) == 1.
    #[test]
    fn lift_single_bits() {
        let v_basis: Vec<BinVec> = vec![0b00101, 0b11000, 0b00010];
        assert_eq!(lift(0, &v_basis), 0);
        assert_eq!(lift(0b001, &v_basis), 0b00101);
        assert_eq!(lift(0b010, &v_basis), 0b11000);
        assert_eq!(lift(0b100, &v_basis), 0b00010);
        // Two bits → XOR of the corresponding rows.
        assert_eq!(lift(0b011, &v_basis), 0b00101 ^ 0b11000);
        assert_eq!(lift(0b111, &v_basis), 0b00101 ^ 0b11000 ^ 0b00010);
    }

    /// `reduce_mod_c(c, _) == 0` for every `c ∈ C`. For `[I_2 | 11]`,
    /// `C = {000, 101, 110, 011}`; the pivots are 0 and 1.
    #[test]
    fn reduce_mod_c_kills_codewords() {
        let rref: Vec<BinVec> = vec![0b101, 0b110]; // pivots 0, 1
        let pivots: Vec<u32> = vec![0, 1];
        for &c in &[0b000u64, 0b101, 0b110, 0b011] {
            assert_eq!(reduce_mod_c(c, &rref, &pivots), 0);
        }
        // A non-codeword reduces to its unique pivot-free coset rep.
        let v = 0b100; // = pivot-1 codeword + bit 2; mod C → bit 2 only
        // Actually 0b100 has bit 2 set, no pivot bits, so already reduced.
        assert_eq!(reduce_mod_c(v, &rref, &pivots), 0b100);
    }

    /// `[8,4]` extended Hamming (self-dual, doubly even) — the canonical
    /// small doubly even code. We hard-code its RREF, dual, and confirm
    /// Q_basis is empty since C = C⊥ ⇒ dim V = n - 2k = 0.
    #[test]
    fn q_basis_extended_hamming_8_4_is_empty() {
        // Extended [8,4] Hamming, self-dual:
        // rows: 1 0 0 0 0 1 1 1
        //       0 1 0 0 1 0 1 1
        //       0 0 1 0 1 1 0 1
        //       0 0 0 1 1 1 1 0
        // Bit 0 = column 0 = LSB.
        // Pivots: 0, 1, 2, 3.
        let rref: Vec<BinVec> = vec![
            0b1110_0001, // c0 = 1 + c5 + c6 + c7
            0b1101_0010, // c1 = 1 + c4 + c6 + c7  (bit 1 is the LSB after?)
            0b1011_0100, // c2 = ...
            0b0111_1000, // c3 = ...
        ];
        let pivots: Vec<u32> = vec![0, 1, 2, 3];
        // For a self-dual code C = C⊥, so dual_basis ⊆ C and every entry
        // reduces to 0; V_basis is empty.
        let dual_basis = rref.clone();
        let (v_basis, pivots_v) = q_basis(&dual_basis, &pivots, &dual_basis, 8);
        assert!(v_basis.is_empty(), "expected empty V, got {v_basis:?}");
        assert!(pivots_v.is_empty(), "expected empty pivots_V, got {pivots_v:?}");
    }

    /// `aut_image_on_q` for an identity automorphism preserves V-basis identity.
    #[test]
    fn aut_image_of_identity_is_identity() {
        // Same V_basis as the round-trip test.
        let v_basis: Vec<BinVec> = vec![0b00001, 0b00100, 0b01000];
        let pivots_v: Vec<u32> = vec![0, 2, 3];
        // C is trivial (zero code) for simplicity: empty rref, empty pivots.
        let rref: Vec<BinVec> = vec![];
        let pivots: Vec<u32> = vec![];
        // identity permutation on 5 columns
        let identity_perm: ColPerm = (0..5u32).collect();
        let aut_generators = vec![identity_perm];
        let images = aut_image_on_q(&aut_generators, &rref, &pivots, &v_basis, &pivots_v);
        assert_eq!(images.len(), 1);
        // Image of e_0, e_1, e_2 under identity is the L×L identity.
        let identity_mat: Mat = vec![0b001, 0b010, 0b100];
        assert_eq!(images[0], identity_mat);
    }

    /// Swap permutation on the two free columns of `[2,1]` code C = {00, 11}.
    /// `V_basis = (0b10,)` (one free column, bit 1), `pivots_V = (1,)`.
    /// Swapping columns 0 and 1 in F_2^2 should send `b = 0b10` to `0b01`,
    /// which reduces mod C (XOR with `0b11`) to `0b10`. So σ_Q is the
    /// identity on Q.
    #[test]
    fn aut_image_swap_columns_of_repetition_code() {
        let rref: Vec<BinVec> = vec![0b11]; // C = ⟨11⟩
        let pivots: Vec<u32> = vec![0];
        let v_basis: Vec<BinVec> = vec![0b10]; // V = ⟨e_1⟩
        let pivots_v: Vec<u32> = vec![1];
        // swap columns 0 and 1
        let swap: ColPerm = vec![1, 0];
        let images = aut_image_on_q(&vec![swap], &rref, &pivots, &v_basis, &pivots_v);
        assert_eq!(images, vec![vec![0b1]]); // 1x1 identity
    }
}
